chain/chain/src/chain.rs

```diff
@@ -1874,80 +1957,106 @@ impl Chain {
         let (apply_chunk_work, block_preprocess_info) = preprocess_res;
         let block = block.into_inner();
         let block_hash = *block.hash();
+        let apply_chunks_done_marker = block_preprocess_info.apply_chunks_done.clone();
         self.blocks_in_processing.add(block, block_preprocess_info)?;
 
         // 2) apply chunks, this is where the transactions and receipts are processed. At this step,
         //    there still no change to ChainStore.
-        self.async_apply_chunks(block_hash, apply_chunk_work);
+        self.async_apply_chunks(
+            block_hash,
+            apply_chunk_work,
+            apply_chunks_done_marker,
+            apply_chunks_done_callback.clone(),
+        );
 
-        // TODO: this will be moved out of process_block once we make do_apply_chunks async
-        self.postprocess_ready_blocks(me, &mut block_processing_artifact.accepted_blocks)
+        Ok(())
     }
 
+    /// Applying chunks async by starting the work at the rayon thread pool
+    /// `apply_chunks_done_marker`: a marker that will be set to true once applying chunks is finished
+    /// `apply_chunks_done_callback`: a callback that will be called once applying chunks is finished
     fn async_apply_chunks(
         &self,
         block_hash: CryptoHash,
         work: Vec<Box<dyn FnOnce(&Span) -> Result<ApplyChunkResult, Error> + Send>>,
+        apply_chunks_done_marker: Arc<OnceCell<bool>>,
+        apply_chunks_done_callback: ApplyChunkCallback,
     ) {
         let sc = self.apply_chunks_sender.clone();
-        // TODO: uncomment this line when we are ready to move apply_chunks async
-        //rayon::spawn(move || {
-        let res = do_apply_chunks(work);
-        // we can safely unwrap here because error means the receiver is deallocated, which means
-        // the chain thread is shut down, and the node already crashed
-        sc.send((block_hash, res)).unwrap();
-        //});
+        rayon::spawn(move || {
+            let res = do_apply_chunks(work);
+            // we can safely unwrap here because error means the receiver is deallocated, which means
+            // the chain thread is shut down, and the node already crashed
+            sc.send((block_hash.clone(), res)).unwrap();
+            if let Err(_) = apply_chunks_done_marker.set(true) {
+                // This should never happen, if it does, it means there is a bug in our code.
+                error!(target:"chain", "apply chunks are called twice for block {:?}", block_hash);
+                debug_assert!(false);
+            }
+            apply_chunks_done_callback(block_hash);
```

Comment:
I don't quite like that we use three separate completion signals for the same thing:

* `once_cell`
* send over the channel
* callback invocation

At the same time, I don't see any obvious simplifications here. 

Long term, I think it should be possible to get by only with a channel send here, after a) we remove actix and the need for hacky priority b) we re-shuffle code in chain and chain actor to make sure that only actor has concurrent parts. That `b)` semes like a right separation to me: chain manages atomic state transitions, while chain actor manages concurrent processes. But that's for another refactor

Reply:
Actually I realized that I can keep only one of once_cell and and channel. I initially implemented this once_cell mechanism outside of Client in test, then realized that the code is cleaner if I move it inside client. I'll probably remove the channel implementation and just use once_cell (to check if new blocks are ready, I can check if the once_cell is set). What do you think?

Reply:
If that works, sgtm! I have doubts that it'll work though. If I understand the idea correctly, I think you want to change the cell to hold `BlockApplyChunksResult`: `OnceCell<BlockApplyChunksResult>`. That way you indeed can use `onec_cell` to asyncronosly notify that the block is done, together with communicating a reslut. What you wouldn't be able to do is to extract the `BlockApplyChunksResult` out of the cell, you'd have to clone it, which I think is something we'd rather avoid. To extract result, you'd need `&mut OnceCell<T>`, but you are sharing it between threads. 

:thinking: well, I guess there might actually be a super-hacky way to extract results still. Two threads can hold two clones of `Arc<OnecCell<T>>`. Producer thread can `.set` the value and then *drop* the `Arc`. The consumer thread will then be able to do Arc::try_unwrap (https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.try_unwrap) and then extract the value. Not sure I like this though